plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,450 @@
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v2/volumes"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/hypervisors"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/flavors"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/projects"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/services"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const (
+	// plugin is used to identify ourselves in log output
+	plugin = "openstack"
+)
+
+// serviceType is an OpenStack service type
+type serviceType string
+
+const (
+	volumeV2Service serviceType = "volumev2"
+)
+
+// tagMap maps a tag name to value.
+type tagMap map[string]string
+
+// fieldMap maps a field to an arbitrary data type.
+type fieldMap map[string]interface{}
+
+// serviceMap maps service id to a Service struct.
+type serviceMap map[string]services.Service
+
+// ContainsService indicates whether a particular service is enabled
+func (s serviceMap) ContainsService(t serviceType) bool {
+	for _, service := range s {
+		if service.Type == string(t) {
+			return true
+		}
+	}
+	return false
+}
+
+// projectMap maps a project id to a Project struct.
+type projectMap map[string]projects.Project
+
+// hypervisorMap maps a hypervisor id a Hypervisor struct.
+type hypervisorMap map[string]hypervisors.Hypervisor
+
+// serverMap maps a server id to a Server struct.
+type serverMap map[string]servers.Server
+
+// flavorMap maps a flavor id to a Flavor struct.
+type flavorMap map[string]flavors.Flavor
+
+// volume is a structure used to unmarshal raw JSON from the API into.
+type volume struct {
+	volumes.Volume
+	volumetenants.VolumeTenantExt
+}
+
+// volumeMap maps a volume id to a volume struct.
+type volumeMap map[string]volume
+
+// storagePoolMap maps a storage pool name to a StoragePool struct.
+type storagePoolMap map[string]schedulerstats.StoragePool
+
+// OpenStack is the main structure associated with a collection instance.
+type OpenStack struct {
+	// Configuration variables
+	IdentityEndpoint string
+	Domain           string
+	Project          string
+	Username         string
+	Password         string
+
+	// Locally cached clients
+	identity *gophercloud.ServiceClient
+	compute  *gophercloud.ServiceClient
+	volume   *gophercloud.ServiceClient
+
+	// Locally cached resources
+	services     serviceMap
+	projects     projectMap
+	hypervisors  hypervisorMap
+	flavors      flavorMap
+	servers      serverMap
+	volumes      volumeMap
+	storagePools storagePoolMap
+}
+
+// Description returns a description string of the input plugin and implements
+// the Input interface.
+func (o *OpenStack) Description() string {
+	return "Collects performance metrics from OpenStack services"
+}
+
+// sampleConfig is a sample configuration file entry.
+var sampleConfig = `
```

Comment:
Can you please comment all options that just specify the default values. The idea is if a user creates a config he/she directly sees the mandatory option to fill-in.